File: framework/java/src/main/java/soda/unittest/validate/StrategyFactory.java
package soda.unittest.validate;

import java.util.List;
import java.util.function.BiPredicate;

public class StrategyFactory {

    public static <T> BiPredicate<List<T>, List<T>> unorderList(ObjectFeature<T> feat) {
        return (List<T> a, List<T> b) -> {
            if (a.size() != b.size()) {
                return false;
            }
            var xmap = new XMap<T,Integer>(feat);
            for (var e : a) {
                xmap.put(e, xmap.getOrDefault(e, 0) + 1);
            }
            for (var e : b) {
                if (!xmap.contains(e)) {
                    return false;
                }
                int c = xmap.get(e) - 1;
                if (c == 0) {
                    xmap.remove(e);
                } else {
                    xmap.put(e, c);
                }
            }
            return true;
        };
    }

    public static <T> BiPredicate<List<T>, List<T>> list(ObjectFeature<T> feat) {
        return (List<T> a, List<T> b) -> {
            if (a.size() != b.size()) {
                return false;
            }
            for (int i = 0; i < a.size(); ++i) {
                if (!feat.isEqual(a.get(i), b.get(i))) {
                    return false;
                }
            }
            return true;
        };
    }

}

File: framework/scala/src/main/scala/soda/scala/unittest/StrategyFactory.scala
package soda.scala.unittest

object StrategyFactory {

  def unorderList[T](feat: ObjectFeature[T]): (List[T], List[T]) => Boolean = {
    def func(a: List[T], b: List[T]): Boolean = {
      if (a.size != b.size) {
        return false
      }
      val xmap = new XMap[T,Int](feat)
      for (e <- a) {
        xmap.put(e, xmap.getOrElse(e, 0) + 1)
      }
      for (e <- b) {
        if (!xmap.contains(e)) {
          return false
        }
        val c = xmap.get(e).get - 1
        if (c == 0) {
          xmap.remove(e)
        } else {
          xmap.put(e, c)
        }
      }
      true
    }
    func
  }

  def list[T](feat: ObjectFeature[T]): (List[T], List[T]) => Boolean = {
    def func(a: List[T], b: List[T]): Boolean = {
      a.size == b.size && a.zip(b).forall(x => feat.isEqual(x._1, x._2))
    }
    func
  }

}
